torch/_export/passes/replace_set_grad_with_hop_pass.py

```diff
@@ -160,7 +160,58 @@ def _maybe_inline_or_replace_with_hop(node: torch.fx.Node):
     return gm
 
 
-def replace_set_grad_with_hop_pass(gm: torch.fx.GraphModule):
+def replace_set_grad_with_hop_pass(gm: torch.fx.GraphModule, graph_signature):
+    original_output_nodes = ()
+    for node in gm.graph.nodes:
+        if node.op == "output":
+            original_output_nodes = node.args
+
+    gm = _replace_set_grad_with_hop_pass(gm)
+
+    updated_output_nodes = ()
+    for node in gm.graph.nodes:
+        if node.op == "output":
+            updated_output_nodes = node.args
+
+    # It could be updated because replace_set_grad_with_hop_pass adds new
+    # getitem nodes into the graph.
+    assert len(original_output_nodes) == len(updated_output_nodes)
+    old_output_name_to_new_output_name = {}
+    for k, v in zip(*original_output_nodes, *updated_output_nodes):
+        if isinstance(k, torch.fx.Node) and isinstance(k, torch.fx.Node):
+            old_output_name_to_new_output_name[k.name] = v.name
+        # If there is constant return in the end, it should be true for updated outputs too
+        else:
+            constant_types = (int, float, str, type(None))
+            assert k == v
+            assert isinstance(k, constant_types)
+
+    buffers_to_mutate_copy = graph_signature.buffers_to_mutate.copy()
+    user_inputs_to_mutate_copy = graph_signature.user_inputs_to_mutate.copy()
+    for k in old_output_name_to_new_output_name:
+        if k in graph_signature.buffers_to_mutate:
+            graph_signature.buffers_to_mutate[
+                old_output_name_to_new_output_name[k]
+            ] = buffers_to_mutate_copy[k]
+            if k not in old_output_name_to_new_output_name.values():
+                del graph_signature.buffers_to_mutate[k]
+
+        if k in graph_signature.user_inputs_to_mutate:
+            graph_signature.user_inputs_to_mutate[
+                old_output_name_to_new_output_name[k]
+            ] = user_inputs_to_mutate_copy[k]
+            if k not in old_output_name_to_new_output_name.values():
+                del graph_signature.user_inputs_to_mutate[k]
+
+    for i, k in enumerate(graph_signature.user_outputs):
+        if k in old_output_name_to_new_output_name:
+            new_k = old_output_name_to_new_output_name[k]
+            graph_signature.user_outputs[i] = new_k
```

Comment:
I think you can get away these regeneration code with:
```
with graph_module._set_replace_hook(ep.graph_signature.get_replace_hook():
    _replace_set_grad_with_hop_pass(gm)
```

Reply:
Hmm we actually cannot use ep here because this is before we create ep. We need it before ep because ep verifier won't pass as we have non-functional ops in the graph at this point.

Reply:
qq: is ep.graph_signature the same as the type as the graph signature here? If so, we could still use the method?

Reply:
discussed offline with @tugsbayasgalan, we should still use something more structural here. will follow up